mapproxy/cache/s3.py

```diff
@@ -0,0 +1,168 @@
+
+    def tile_key(self, tile):
+        return self._tile_location(tile, self.base_path, self.file_ext).lstrip('/')
+
+    def conn(self):
+        if boto3 is None:
+            raise ImportError("S3 Cache requires 'boto3' package.")
+
+        try:
+            return s3_session().client("s3")
+        except Exception as e:
+            raise S3ConnectionError('Error during connection %s' % e)
+
+    def load_tile_metadata(self, tile):
+        if tile.timestamp:
+            return
+        self.is_cached(tile)
+
+    def _set_metadata(self, response, tile):
+        if 'LastModified' in response:
+            tile.timestamp = float(response['LastModified'].strftime('%s'))
+        if 'ContentLength' in response:
+            tile.size = response['ContentLength']
+
+    def is_cached(self, tile):
+        if tile.is_missing():
+            key = self.tile_key(tile)
+            try:
+                r = self.conn().head_object(Bucket=self.bucket_name, Key=key)
+                self._set_metadata(r, tile)
+            except botocore.exceptions.ClientError as e:
+                if e.response['Error']['Code'] in ('404', 'NoSuchKey'):
+                    return False
+                raise
+
+        return True
+
+    def load_tiles(self, tiles, with_metadata=True):
+        p = async.Pool(min(4, len(tiles)))
+        return all(p.map(self.load_tile, tiles))
+
+    def load_tile(self, tile, with_metadata=True):
+        if not tile.is_missing():
+            return True
+
+        key = self.tile_key(tile)
+        log.debug('S3:load_tile, key: %s' % key)
+
+        try:
+            r  = self.conn().get_object(Bucket=self.bucket_name, Key=key)
+            self._set_metadata(r, tile)
+            tile.source = ImageSource(r['Body'])
+        except botocore.exceptions.ClientError as e:
+            error = e.response.get('Errors', e.response)['Error'] # moto get_object can return Error wrapped in Errors...
+            if error['Code'] in ('404', 'NoSuchKey'):
+                return False
+            raise
+
+        return True
+
+    def remove_tile(self, tile):
+        key = self.tile_key(tile)
+        log.debug('remove_tile, key: %s' % key)
+        self.conn().delete_object(Bucket=self.bucket_name, Key=key)
+
+    def store_tiles(self, tiles):
+        p = async.Pool(min(4, len(tiles)))
+        p.map(self.store_tile, tiles)
+
+    def store_tile(self, tile):
+        if tile.stored:
+            return
+
+        key = self.tile_key(tile)
+        log.debug('S3: store_tile, key: %s' % key)
+
+        extra_args = {}
+        if self.file_ext in ('jpeg', 'png'):
+            extra_args['ContentType'] = 'image/' + self.file_ext
+        with tile_buffer(tile) as buf:
+            self.conn().upload_fileobj(
+                NopCloser(buf), # upload_fileobj closes buf, wrap in NopCloser
+                self.bucket_name,
+                key,
+                ExtraArgs=extra_args)
```

Comment:
MapProxy doesn't create buckets on its own. The user can set the default storage class when creating the bucket on the AWS dashboard, right?